Fix base URL bug

lukasgeiter opened the pull request:
_This solves the issue #8441_

It might not be considered as a real bug, but the Symfony Request makes a mistake when determining the base URL of a request.
If the application is running under domain root, following URLs will all produce the same path info:

```
http://laravel.app/
http://laravel.app/foo/bar/index.php
```

`foo/bar` can be replaced with anything, the page will still work and **not** return a 404.

For a live example hit: http://laravel.com/foo/index.php/docs/5.0/

This PR fixes that behavior but I'd really like some feedback from you guys:
1. Is it OK to fix this just within Laravel or should it be fixed in the Symfony Request class?
2. I'm unsure if this covers all (necessary) cases. Especially since the original `prepareBaseUrl()` is a lot more complicated. What do you think?

lukasgeiter committed:
Fix base URL bug

GrahamCampbell commented:
Please fix all the cs issues.

GrahamCampbell commented:
Does this break things for people running laravel in a folder?

lukasgeiter commented:
Sorry about the code style. Should be good now...

I does not break things for running it in sub directory. (At least it didn't for me)

GrahamCampbell commented on src/Illuminate/Http/Request.php:
```diff
@@ -630,6 +630,22 @@ public static function createFromBase(SymfonyRequest $request)
+	protected function prepareBaseUrl()
+	{
+		$scriptName = $this->server->get('SCRIPT_NAME');
+		if(starts_with(ltrim($this->server->get('REQUEST_URI'), '/'), ltrim($scriptName, '/')))
```
missing space

lukasgeiter replied:
Where exactly? Are there any more specific style guidelines regarding spacing? (other than PSR-1 and that bit in the [docs](http://laravel.com/docs/5.0/contributions#coding-style))

GrahamCampbell replied:
before the opening bracket just after the if.

GrahamCampbell commented on src/Illuminate/Http/Request.php:
```diff
@@ -630,6 +630,22 @@ public static function createFromBase(SymfonyRequest $request)
+			// that the filename is used in the URI and use that as base URL
+			return trim($scriptName, '/');
+		}
+		// otherwise we'll just use the directory path relative to the document root
```
missing new line

GrahamCampbell commented on src/Illuminate/Http/Request.php:
```diff
@@ -630,6 +630,22 @@ public static function createFromBase(SymfonyRequest $request)
 	 */
+	protected function prepareBaseUrl()
+	{
+		$scriptName = $this->server->get('SCRIPT_NAME');
```
you need a blank line following this assignment before the control statement.

lukasgeiter commented:
I've fixed all the coding style issues (I hope) and also fixed a little bug for generating URLs when running under root on windows (strangely `dirname()` will return a backslash which needs to be trimmed)

montogeek commented:
Thanks!

GrahamCampbell commented:
Please squash to one commit.

GrahamCampbell commented:
I actually can't replicate this bug anyway. StyleCI runs laravel 5.0, and https://styleci.io/123/donate shows 404 correctly. It looks like it's caused by server missconfiguration, rather than a bug with laravel.

lukasgeiter commented:
You have to actually put `index.php` in the URL. Like https://styleci.io/123/index.php/donate

GrahamCampbell commented:
Oh right. My bad.

taylorotwell commented:
Then send a PR to Symfony.

lukasgeiter commented:
Hmm. In an effort to do that I just discovered that the symfony site itself handles this case correctly.
http://symfony.com/foo/app.php/jobs shows a 404 page like it should.

GrahamCampbell commented:
symfony.com is running drupal though, so it's not totally using the same routing system.

lukasgeiter commented:
Ahh well that's an explanation ;)
Do you know any symfony sites?

PatrickBauer commented:
symfony.com is not running on drupal. At least there is no indicator for this. Why wouldn't it run on symfony?

aequasi commented:
HEAVILY NSFW

Only site i can think of http://www.youporn.com/app/app.php.

But it does do work as this bug suggests.

lukasgeiter commented:
@aequasi Haha I definitely won't use that example in a bug report / PR...

aequasi commented:
Yeah... Im trying to find another one... Also @PatrickBauer, because symfony would be overkill for symfony.com

javiereguiluz commented:
Sorry to interrupt your discussion, but regarding this comment:

> symfony.com is running drupal though, so it's not totally using the same routing system.

I'm afraid that's not true. I'm the guy in charge of symfony.com development, so I can assure you that the site is developed using the latest stable version of the Symfony full-stack framework.

lukasgeiter commented:
@javiereguiluz You're not interrupting at all. Thanks for clearing that up.

I guess my next step will be setting up a symfony application myself to do some testing...

aequasi commented:
set up a quick app on Heroku for the PR, using the demo app. Should be pretty easy

lukasgeiter commented:
@aequasi Sure, just thought it would be nice to demonstrate the bug on a big site...

lancepioch commented:
@lukasgeiter Here's a fun [example](http://laravel.com/Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate.Loremipsumdolorsitamet,consecteturadipiscingelit.Phasellustristiqueenimquisornarecursus.Sednecdignissimturpis,ultriciesefficiturnunc.Inegetnunclacus.Praesentultricesexquisestviverra,euconsequatnequealiquam.Aliquamhendreritauguevitaesagittisfaucibus.Morbiconsequatpharetraarcuatpharetra.Vivamusquismattisodio,acvestibulumarcu.Quisquesitamettempuspurus.Aliquameuismod,telluseutempusposuere,duimassasuscipitsapien,vitaeegestasdolorlectusnecleo.Suspendissepotenti.Pellentesquerutrumultricesdolor.Pellentesquecongueurnaeuexeuismod,aliquamrhoncusnullamolestie.Nullaaenimegetfelisaliquetblanditidnonnibh.Crassapieneros,convallisnonbibendumac,portaegetpurus.Aliquamsitametloremsitametloremlaoreettincidunt.Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate..Vestibulumrhoncusbibendumenimsitametvulputate./index.php/).

aequasi commented:
what the hell?

garygreen commented:
@taylorotwell 

http://laravel.com/docs/5.0/yeah/this/might/be/a/symphony/bug/but/meh/index.php/docs/5.0/controllers

Maybe you should think about shortening Larvel's urls? ;D

garygreen commented:
I did a issue for this on Symphony, seems there is a PR to fix the issue hopefully will get merged. Haven't tested if it fixes the issue yet symfony/symfony#13617 pr: symfony/symfony#14335

lukasgeiter commented:
@garygreen I've just tried the PR and it didn't change anything...

I also found a live example: http://cmf.symfony.com/foo/bar/app.php/news

lukasgeiter commented:
Created a PR: symfony/symfony#14468

Let's see how that goes...
